client/src/NoteForm.js

```diff
@@ -45,14 +57,16 @@ class NoteForm extends React.Component {
 
   handleTitleChange() {
     this.shouldRerender = true;
-    this.state.note.title = this.$title.val();
-    this.setState({ note: this.state.note }, this.handleChange);
+    const note = this.state.note;
+    note.title = this.$title.val();
+    this.setState({ note: note }, this.handleChange);
   }
 
   handleContentChange() {
     this.shouldRerender = false;
-    this.state.note.content = this.$content.html();
-    this.setState({ note: this.state.note }, this.handleChange);
+    const note = this.state.note;
+    note.content = this.$content.html();
+    this.setState({ note: note }, this.handleChange);
```

Comment:
you can use the more "immutable" way:

```
this.setState({
 note: {
    ...this.state.note,
    content: this.$content.html()
}
}, this.handleChange)
```

reason: `setState` might shallowly check whether `note` has changed. In your case it hasn't, because its still the same instance, whereas with the spreading approach, you creating a new object.

TbH, i don't now whether `setState` does a shallow-compare or not, but i other libraries like `redux` do that and would not trigger an update in this case.

Reply:
The problem with this is a bit that the spread note needs to be wrapped again as a `Note` as otherwise the expected functions like `toJson` aren't available. Being able to construct a new note from existing attributes complexes the model class, it gets messy pretty fast (also due to the use of getters/setters, because the properties of the object have a different name when spread. Should I refrain from using getters?).

Reply:
Its actually not that common to have class instances as state objects (with getters and stuff) in react. State is usually a simple data struct(ure), so you tend to have helper functions to format or calculate values (or so called "selectors", which return a value (or calculated value) base on your whole `State`). This is derived from functional languages, where you also have simple types without methods and treat all your variables and objects as immutable (admittedly Javascript lacks the full elegancy of real functional languages, so the advantages of this paradigm does not shine through so much). 

Tbh. i don't know whether `setState` actually cares if the object reference changes or not, but i know that libraries like redux rely on that. So concerning state, I follow the rule: if an object changes, you create a new one.

Reply:
> so you tend to have helper functions to format or calculate values (or so called "selectors", which return a value (or calculated value) base on your whole State)

And where do you usually put those? Service classes? Do you have examples of how to structure code for this?

Reply:
> Its actually not that common to have class instances as state objects (with getters and stuff)

So don't you work with classes at all in react? Just plain objects?